### ISSUE TITLE
opty does not accept only one scalar differential equation

### ISSUE THREAD

**Issue by Peter230655** [csu-hmc/opty#255]
Problem 10.7 in Chapter 10 of John T. Betts' _Practical Methods for..._.., 3rd edition, only has one scalar differential equation:
$\dot y = - y^3 + u$, where u is the control.
When I run it, I get this error:
![image](https://github.com/user-attachments/assets/3dc12339-c50c-4fc0-81cf-f16ae64d8cfc)

When I 'artificially' enlarge the system:
$uy = \dfrac{d}{dt} y$
$uy = - y^3 + u$
all works fine and it find the solution.
This is definitely not a big thing, it is probably a rare case that there is only one scalar DE, but I wanted to point it out.

**Comment by moorepants** [comment]
Yes, opty assumes you are working with a physics system that has a mass and forces, thus you always get at least one second order differential equation.

**Comment by Peter230655** [comment]
> Yes, opty assumes you are working with a physics system that has a mass and forces, thus you always get at least one second order differential equation.

Makes sense.
Betts gives the dynamic equations as a set of first order DEs, at least in the three examples I have calculated.
In example 10.6 it is a set of 10 first order DEs, which do not look like mechanical equations - yet the results look fine.
So, as long as it is more than one first order DEs in the eoms, opty does not care how they 'originated?

**Comment by moorepants** [comment]
>  So, as long as it is more than one first order DEs in the eoms, opty does not care how they 'originated?

I think that the only assumption is that there are n states and n equations, where n is > 1.

Two things:

- An example in the gallery that shows how to use a single differential equation with your hack is useful as a temporary solution.
- We should relax the assumption opty makes, because we also would like to add extra constraint equations where n states != num constraint equations (e.g. in https://github.com/csu-hmc/opty/issues/156).

**Comment by Peter230655** [comment]
> This issues is sufficient for the request for support for single first order differential equations.

Not sure I understand. Raise a new issue or not?
(Sorry, my command of the English language...)

**Comment by moorepants** [comment]
This is the issue, no new issue is needed.

**Comment by moorepants** [comment]
You could try removing https://github.com/csu-hmc/opty/blob/master/opty/direct_collocation.py#L1184 and just see what happens.

**Comment by Peter230655** [comment]
> You could try removing https://github.com/csu-hmc/opty/blob/master/opty/direct_collocation.py#L1184 and just see what happens.

I will try in the next couple of days and report.

**Comment by Peter230655** [comment]
```
# %%
"""
Hypersensitive Control
======================

This is example 10.7 from Betts' Test Problems.

**States**

- y : state variable
- uy: its speed

**Specifieds**

- u : control variable

Note: the state variable uy is needed because opt currently needs minimum two
differential equations in the equations of motion. Mathematically it is not
needed.

"""

import numpy as np
import sympy as sm
import sympy.physics.mechanics as me
from opty.direct_collocation import Problem
from opty.utils import create_objective_function
import matplotlib.pyplot as plt

# %%
# Equations of motion.

t = me.dynamicsymbols._t
y, uy, u = me.dynamicsymbols('y uy, u')

eom = sm.Matrix([ uy - y.diff(t), -uy - y**3 + u])
sm.pprint(eom)

# %%

def solve_optimization(nodes, tf):
    t0, tf = 0.0, tf
    num_nodes = nodes
    interval_value = (tf - t0)/(num_nodes - 1)

    # Provide some reasonably realistic values for the constants.

    state_symbols = (y, uy)
    specified_symbols = (u,)


    # Specify the objective function and form the gradient.
    obj_func = sm.Integral(y**2 + u**2, t)
    sm.pprint(obj_func)
    obj, obj_grad = create_objective_function(obj_func,
                                          state_symbols,
                                          specified_symbols,
                                          tuple(),
                                          num_nodes,
                                          node_time_interval=interval_value)


    # Specify the symbolic instance constraints, as per the example
    instance_constraints = (
        y.func(t0) - 1,
        y.func(tf) - 1.5,
    )


    # Create the optimization problem and set any options.
    prob = Problem(obj, obj_grad, eom, state_symbols,
               num_nodes, interval_value,
               instance_constraints=instance_constraints,
    )

    prob.add_option('nlp_scaling_method', 'gradient-based')

    # Give some rough estimates for the trajectories.
    initial_guess = np.zeros(prob.num_free)

    # Find the optimal solution.
    solution, info = prob.solve(initial_guess)
    print(info['status_msg'])
    print(f'Objective value achieved: {info['obj_val']:.4f}, as per the book ' +
        f'it is {6.7241}, so the error is: '
        f'{(info['obj_val'] - 6.7241)/6.7241*100:.3f} % ')

    # Plot the optimal state and input trajectories.
    prob.plot_trajectories(solution)

    # Plot the constraint violations.
    prob.plot_constraint_violations(solution)

    # Plot the objective function as a function of optimizer iteration.
    prob.plot_objective_value()

# %%
# As per the example tf = 10000

tf = 10000
num_nodes = 501
solve_optimization(num_nodes, tf)

# %%
# With the value of tf = 10000 above, opty converged to a locally optimal point,
# but the objective value is far from the one given in the book.
# As per the plot of the solution y(t) it seems, that most of the time y(t) = 0,
# only at the very beginning and the very end it is different from 0.
# So, it may make sense to use a smaller tf.
# Also increasing num_nodes may help.

tf = 8.0
num_nodes = 10001
solve_optimization(num_nodes, tf)

# %%

# sphinx_gallery_thumbnail_number = 4

plt.show()
```

**Comment by moorepants** [comment]
You need to open a PR with the new example if you want help with it. Otherwise we can't run it in context or see what errors the CI produces.

**Comment by Peter230655** [comment]
Will do so tomorrow morning.

**Comment by Peter230655** [comment]
I opened PR 256 this morning.

**Comment by Peter230655** [comment]
Problem was the wrong file name. For examples-gallery it must start with plot_...

**Comment by moorepants** [comment]
Reopened, this is still an issue we can solve.

**Comment by Peter230655** [comment]
> Reopened, this is still an issue we can solve.

I thought you set up examples-gallery like this on purpose.

**Comment by moorepants** [comment]
This issue is about opty needing support for one scalar differential equation. When we add that support, we can close this issue.

**Comment by Peter230655** [comment]
> This issue is about opty needing support for one scalar differential equation. When we add that support, we can close this issue.

This is what I get for mixing two things in the same issue! Sorry!

**Comment by Peter230655** [comment]
> > You could try removing https://github.com/csu-hmc/opty/blob/master/opty/direct_collocation.py#L1184 and just see what happens.
> 
> I will try in the next couple of days and report.

Took me a bit longer than a couple of days. :-)
If I remove the two liners you mentioned, and a similar query around line 1598 it seems to work fine with one eom only.
(I tested only one example, example 10.7 from Betts' book.
**Should I make a PR which removes these four lines?**

**Comment by moorepants** [comment]
Yes, please make a PR. Include a unit test that has a simple one equation.

**Comment by Peter230655** [comment]
> Yes, please make a PR. Include a unit test that has a simple one equation.

Will do so. This unit test would be an optimization with only one eom? (I have this, its the one I used to test it.)

**Comment by moorepants** [comment]
I don't think you need to solve the optimization problem, but you need to initialize a Problem with 1 eom() and just test that it evaluates the constraints and jacobian.

**Comment by Peter230655** [comment]
> I don't think you need to solve the optimization problem, but you need to initialize a Problem with 1 eom() and just test that it evaluates the constraints and jacobian.

I check, whether _Problem.jacobian(free)_ returns the correct shape? Free = initial_guess?
Same with _Problem.constraints(free)_, with free = initial_guess which meet the constraints?

Where would I put this test function? Add it to test_collocation.py?

**Comment by Peter230655** [comment]
In the opty documentation release 1.4.0.dev0 on page 129 it says:
jacobian(free)
Returns.....
Return type
ndarray.shape((2*n + q + r + s) * (n*(N-1) + o**)**, )
but in my test it returns
ndarray.shape((2*n + q + r + s) * (n*(N-1)**)** + o, )

Is there a mistale in my test or is this a typo in the documentation)

**Comment by Peter230655** [comment]
opty can accept one (1) DE only, PR #287 should take care of it.
So, I will close this issue and open a new one so the typo in the documentation does not get forgotten.

**Comment by moorepants** [comment]
Fixed by #311
